### PR TITLE
ShanoirUploader: Fix issue#903 loading time import dialog

### PR DESCRIPTION
**Pull request by michaelkain**

### CODE REVIEW

**Comment by michaelkain** [comment]
@jcomedouteau Hi, it would be great if you could maybe do a rapid test for the CsvImport? It might be faster for you to test. I will test more on my side with different user types for the general import with the GUI only. With kind regards, Michael

**Comment by michaelkain** [comment]
@jcomedouteau I am getting all acq equipments at once and then iterate inside to gain time.

File: shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportFromCsvRunner.java [hunk 1]
```diff
@@ -100,20 +100,26 @@ protected Void doInBackground() throws Exception {
...
+			// Iterate over study cards to get equipment + fill study => SC map
 			for (StudyCard studyCard : studyCards) {
-				AcquisitionEquipment acquisitionEquipment;
-				acquisitionEquipment = shanoirUploaderServiceClientNG.findAcquisitionEquipmentById(studyCard.getAcquisitionEquipmentId());
-				if (acquisitionEquipment == null) {
-					throw new ShanoirException("Error while retrieving study cards");
+				Long acquisitionEquipmentIdSC = studyCard.getAcquisitionEquipmentId();
+				for (Iterator<AcquisitionEquipment> acquisitionEquipmentsIt = acquisitionEquipments.iterator(); acquisitionEquipmentsIt.hasNext();) {
+					AcquisitionEquipment acquisitionEquipment = (AcquisitionEquipment) acquisitionEquipmentsIt.next();
+					if (acquisitionEquipment.getId() == acquisitionEquipmentIdSC) {
```
Comment (jcomedouteau) [inline shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportFromCsvRunner.java:117]: This should be an "equals" and not "=="

File: shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java [hunk 2]
```diff
@@ -110,56 +111,75 @@ private void updateImportDialogForMRICenter(final UploadJob uploadJob) {
...
-									studyCard.setCompatible(true);
-									compatibleStudyCard = true;
-								} else {
-									studyCard.setCompatible(false);
+						if (study.getId() == studyCard.getStudyId()) {
+							studyCardsStudy.add(studyCard);
+							Long acquisitionEquipmentId = studyCard.getAcquisitionEquipmentId();
+							for (Iterator<AcquisitionEquipment> acquisitionEquipmentsIt = acquisitionEquipments.iterator(); acquisitionEquipmentsIt.hasNext();) {
+								AcquisitionEquipment acquisitionEquipment = (AcquisitionEquipment) acquisitionEquipmentsIt.next();
+								if (acquisitionEquipment.getId() == acquisitionEquipmentId) {
```
Comment (jcomedouteau) [inline shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java:133]: This should be an "equals" and not "=="

File: shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java [hunk 3]
```diff
@@ -110,56 +111,75 @@ private void updateImportDialogForMRICenter(final UploadJob uploadJob) {
...
-								String manufacturerSC = acquisitionEquipment.getManufacturerModel().getManufacturer().getName();
-								String manufacturerModelNameSC = acquisitionEquipment.getManufacturerModel().getName();
-								if (manufacturerSC.compareToIgnoreCase(manufacturer) == 0
-										&& manufacturerModelNameSC.compareToIgnoreCase(manufacturerModelName) == 0
-										&& acquisitionEquipment.getSerialNumber().compareToIgnoreCase(deviceSerialNumber) == 0) {
-									studyCard.setCompatible(true);
-									compatibleStudyCard = true;
-								} else {
-									studyCard.setCompatible(false);
+						if (study.getId() == studyCard.getStudyId()) {
```
Comment (jcomedouteau) [inline shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java:128]: This should be an "equals" and not "=="

**Comment by jcomedouteau** [comment]
I'm testing it

**Comment by jcomedouteau** [comment]
After my test I feel it's really faster:
2021-04-13 11:15:03 INFO  [org.shanoir.uploader.service.rest.ShanoirUploaderServiceClientNG] [213] findStudiesNamesAndCenters: 1747ms
2021-04-13 11:15:05 INFO  [org.shanoir.uploader.service.rest.ShanoirUploaderServiceClientNG] [231] findStudyCardsByStudyIds: 1390ms

**Comment by michaelkain** [comment]
I have finished the tests on my side too, please find my test report:
Test report, ShUp 2021-04-13, after merge of CsvImport
+ change on loading time of ImportDialog, test imports

- Test 1: with user type User, sample1 used, study TEST_IMPORT
-- Create new subject, new exam, with one serie WebGUI
--- serie T2Star selected, Peter Parker 01.01.1901 -> copy worked, OK
--- ImportDialog open rapidly: 3 studies available?
--- all acq equipments accessed, same for admin role, so no difference -> OK
2021-04-13 10:56:19 INFO  [org.shanoir.uploader.service.rest.ShanoirUploaderServiceClientNG] [213] findStudiesNamesAndCenters: 191ms
2021-04-13 10:56:19 INFO  [org.shanoir.uploader.action.ImportDialogOpenerNG] [114] getStudiesWithStudyCards: 3 studies found.
2021-04-13 10:56:19 INFO  [org.shanoir.uploader.service.rest.ShanoirUploaderServiceClientNG] [335] findAcquisitionEquipments: 272ms
2021-04-13 10:56:20 INFO  [org.shanoir.uploader.action.ImportDialogOpenerNG] [117] findAcquisitionEquipments: 323 equipments found.
2021-04-13 10:56:21 INFO  [org.shanoir.uploader.service.rest.ShanoirUploaderServiceClientNG] [231] findStudyCardsByStudyIds: 1461ms
2021-04-13 10:56:22 INFO  [org.shanoir.uploader.action.ImportDialogOpenerNG] [119] getAllStudyCards: 33 studycards found.
--- new subject to be created automatically into TEST_IMPORT study + new exam created
with name 2021-04-13
--- Import: finished -> OK!
--- Check with WebGUI: 5 datasets generated, dicom+nifti download + preview OK!

-- Test 2: Use existing subject, other study used! Cohort Rennes, new exam, with one entire exam WebGUI
--- entire exam selected
--- OK! existing subject found with name 0080090
--- proposes creation of new exam, even if existing one with same date
--- import started, as big exam, waiting for end of import job
--- OK! now working, only took some time to really finish the import!!!
--- great, now all series are present too

-- Test 3: Use existing subject 0080090, TEST_IMPORT, existing exam same date, with one serie T1
--- OK! now working

File: shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportFromCsvRunner.java (hunk 1 above)
Comment (michaelkain) [reply shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportFromCsvRunner.java:117]: done

File: shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java (hunk 2 above)
Comment (michaelkain) [reply shanoir-uploader/src/main/java/org/shanoir/uploader/action/ImportDialogOpenerNG.java:133]: done

**Comment by michaelkain** [comment]
@jcomedouteau thank you for your tests!

**Comment by michaelkain** [comment]
@jcomedouteau Hi Jean-Come, could you approve the PR?

**Review by jcomedouteau — APPROVED** [review]

**Pull request merged**